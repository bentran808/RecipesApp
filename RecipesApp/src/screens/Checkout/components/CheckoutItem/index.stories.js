import { storiesOf } from '@storybook/react-native';
import { recipe } from 'mocks';
import React from 'react';
import CenterView from 'storybook/stories/CenterView';
import CheckoutItem from '.';

storiesOf('Checkout Item', module)
  .addDecorator((getStory) => <CenterView>{getStory()}</CenterView>)
  .add('default', () => (
    <CheckoutItem
      item={{
        item: recipe,
        quantity: 1,
        inCart: true
      }}
    />
  ));
