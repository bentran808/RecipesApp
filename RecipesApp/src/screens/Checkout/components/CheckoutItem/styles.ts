import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingVertical: 10,
    paddingHorizontal: 20,
    width: '100%'
  },
  title: {
    fontWeight: 'bold'
  },
  subTitle: {
    fontSize: 12,
    color: 'gray'
  },
  quantityWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: 150
  },
  quantityText: {
    borderWidth: 1,
    borderColor: 'red',
    borderRadius: 5,
    color: 'red',
    paddingVertical: 5,
    paddingHorizontal: 30
  }
});

export default styles;
