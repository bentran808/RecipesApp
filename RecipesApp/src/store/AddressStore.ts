import { toJS } from 'mobx';
import { types, destroy, SnapshotOut, getRoot, applySnapshot } from 'mobx-state-tree';
import { RootStore } from 'store/store';

const AddressEntry = types
  .model('AddressEntry', {
    id: types.optional(types.identifierNumber, 0),
    type: types.optional(types.string, ''),
    address: types.optional(types.string, '')
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).address.removeItem(self);
    },
    edit(type: string, address: string) {
      self.type = type;
      self.address = address;
    }
  }));

export type AddressModel = SnapshotOut<typeof AddressEntry>;

const AddressStore = types
  .model('', {
    items: types.array(AddressEntry),
    itemUsing: types.optional(AddressEntry, {})
  })
  .volatile(() => ({
    address: ''
  }))
  .actions((self) => ({
    setAddress(address: string) {
      self.address = address;
    },
    addNewItem(type: string, address: string) {
      const id = self.items.reduce((maxId, item) => Math.max(item.id, maxId), -1) + 1;
      self.items.push({ id, type, address });
    },
    removeItem(item: AddressModel) {
      destroy(item);
    },
    existAddress(address?: AddressModel) {
      return self.items.find((item) => item.id === address?.id);
    },
    setItemUsing(item: AddressModel) {
      applySnapshot(self, {
        ...self,
        itemUsing: item
      });
    }
  }))
  .views((self) => ({
    get addresses() {
      return toJS(self.items);
    }
  }));

export default AddressStore;
