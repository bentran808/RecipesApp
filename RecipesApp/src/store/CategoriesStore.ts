import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';
import { transformCategories } from 'utils';
import { RecipesEntry } from './RecipesStore';

export const CategoriesEntry = types.model('CategoriesEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string,
  recipes: types.maybe(types.array(RecipesEntry))
});

export type CategoryModel = SnapshotOut<typeof CategoriesEntry>;

const CategoriesStore = types
  .model('CategoriesStore', {
    lists: types.optional(types.array(CategoriesEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get categories() {
      return self.lists;
    },
    get categoriesJS() {
      return transformCategories(toJS(this.categories));
    }
  }))
  .actions((self) => ({
    fetchCategories: flow(function* fetchCategories() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchCategoriesRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        console.log(error);
        self.state = 'error';
      }
    })
  }));

export default CategoriesStore;
