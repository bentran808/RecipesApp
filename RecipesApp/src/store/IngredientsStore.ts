import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';

export const IngredientsEntry = types.model('IngredientsEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string
});

export type IngredientModel = SnapshotOut<typeof IngredientsEntry>;

const IngredientsStore = types
  .model('IngredientsStore', {
    lists: types.optional(types.array(IngredientsEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get ingredients() {
      return self.lists;
    },
    get ingredientsJS() {
      return toJS(this.ingredients);
    }
  }))
  .actions((self) => ({
    fetchIngredients: flow(function* fetchIngredients() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchIngredientsRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        self.state = 'error';
      }
    })
  }));

export default IngredientsStore;
