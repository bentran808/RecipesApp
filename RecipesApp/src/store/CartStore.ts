import { toJS } from 'mobx';
import { applySnapshot, destroy, getRoot, SnapshotOut, types } from 'mobx-state-tree';
import { RecipeModel, RecipesEntry } from 'store/RecipesStore';
import { RootStore } from 'store/store';

export const CartEntry = types
  .model('CartEntry', {
    item: RecipesEntry,
    quantity: types.optional(types.number, 1),
    inCart: types.optional(types.boolean, true)
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).cart.removeItem(self);
    },
    increase(quantity?: number) {
      self.quantity = quantity ? self.quantity + quantity : ++self.quantity;
    },
    decrease() {
      self.quantity === 1 ? this.remove() : --self.quantity;
    },
    toggle() {
      self.inCart = !self.inCart;
    }
  }));

const BillEntry = types.model('BillEntry', {
  type: types.optional(types.string, ''),
  name: types.optional(types.string, ''),
  code: types.optional(types.string, ''),
  price: types.optional(types.number, 0)
});

const OrderedEntry = types.model('OrderedEntry', {
  items: types.array(
    types.model({
      name: types.optional(types.string, ''),
      quantity: types.optional(types.number, 0)
    })
  ),
  total: types.number,
  createdAt: types.string
});

export type CartModel = SnapshotOut<typeof CartEntry>;
export type BillModel = SnapshotOut<typeof BillEntry>;
export type OrderedModel = SnapshotOut<typeof OrderedEntry>;

const CartStore = types
  .model('CartStore', {
    items: types.array(CartEntry),
    billItems: types.array(BillEntry),
    orderedItems: types.array(OrderedEntry)
  })
  .views((self) => ({
    get recipes() {
      return toJS(self.items);
    },
    get recipesInCart() {
      return this.recipes.filter((item) => item.inCart);
    },
    get inCartCount() {
      return this.recipesInCart.length;
    },
    get total() {
      return this.recipesInCart.reduce((sum, item) => sum + item.quantity * 20, 0);
    },
    get billDetails() {
      return toJS(self.billItems);
    },
    get toPay() {
      return this.billDetails.reduce(
        (sum, item) => (item.type === 'discount' ? sum - item.price : sum + item.price),
        0
      );
    },
    get orderedList() {
      return toJS(self.orderedItems);
    }
  }))
  .volatile(() => ({
    discountInput: '',
    invalidCoupon: false
  }))
  .actions((self) => ({
    addToCart(item: RecipeModel, quantity?: number) {
      const itemInCart = this.hasInCart(item);
      itemInCart ? itemInCart.increase(quantity) : self.items.push({ item, quantity });
    },
    removeItem(item: CartModel) {
      destroy(item);
    },
    hasInCart(recipe: RecipeModel) {
      return self.items.find((item) => item.item.id === recipe.id);
    },
    setDiscountInput(text: string) {
      if (!text) {
        self.invalidCoupon = false;
      }
      self.discountInput = text;
    },
    applyCoupon(item: BillModel) {
      const existItem = self.billItems.findIndex((bill) => bill.type === 'discount');
      if (existItem > -1) {
        self.billItems.splice(existItem, 1, item);
      } else {
        self.billItems.push(item);
      }
    },
    payment(order: OrderedModel) {
      self.orderedItems.unshift(order);
      applySnapshot(self, {
        ...self,
        items: [],
        billItems: []
      });
    },
    isInvalidCoupon(status: boolean) {
      self.invalidCoupon = status;
    }
  }));

export default CartStore;
