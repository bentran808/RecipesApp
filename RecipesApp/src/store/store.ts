import { Instance, types } from 'mobx-state-tree';
import RecipesStore from 'store/RecipesStore';
import CategoriesStore from './CategoriesStore';

const RootStore = types.model('RootStore', {
  categories: CategoriesStore,
  recipes: RecipesStore
});

const store = RootStore.create({
    categories: {},
    recipes: {}
});

export type RootStoreModel = Instance<typeof RootStore>;

export default store;
