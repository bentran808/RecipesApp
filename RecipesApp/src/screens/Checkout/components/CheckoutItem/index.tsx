import React from 'react';
import { Text, View } from 'react-native';
import { CartModel } from 'store/CartStore';
import styles from './styles';

type Props = {
  item: CartModel;
};

const CheckoutItem = ({ item }: Props) => {
  return (
    <View style={styles.container}>
      <View>
        <Text style={styles.title}>{item.item.title}</Text>
        <Text style={styles.subTitle}>Regular</Text>
      </View>
      <View style={styles.quantityWrapper}>
        <Text style={styles.quantityText}>{item.quantity}</Text>
        <Text>${item.quantity * 20}</Text>
      </View>
    </View>
  );
};

export default React.memo(CheckoutItem);
