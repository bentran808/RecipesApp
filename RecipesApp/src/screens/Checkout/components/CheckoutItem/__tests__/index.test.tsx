import { StoreProvider } from 'context';
import { recipe } from 'mocks';
import React from 'react';
import renderer from 'react-test-renderer';
import store from 'store/store';
import CheckoutItem from '..';

describe('Checkout Item Component', () => {
  const props = {
    item: {
      item: recipe,
      quantity: 1,
      inCart: true
    }
  };

  test('should render correctly', () => {
    const tree = renderer
      .create(
        <StoreProvider value={store}>
          <CheckoutItem {...props} />
        </StoreProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
