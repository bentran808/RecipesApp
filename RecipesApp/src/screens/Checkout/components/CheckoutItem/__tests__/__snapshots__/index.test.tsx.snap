// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout Item Component should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 20,
      "paddingVertical": 10,
      "width": "100%",
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Oatmeal Cookies
    </Text>
    <Text
      style={
        Object {
          "color": "gray",
          "fontSize": 12,
        }
      }
    >
      Regular
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 150,
      }
    }
  >
    <Text
      style={
        Object {
          "borderColor": "red",
          "borderRadius": 5,
          "borderWidth": 1,
          "color": "red",
          "paddingHorizontal": 30,
          "paddingVertical": 5,
        }
      }
    >
      1
    </Text>
    <Text>
      $
      20
    </Text>
  </View>
</View>
`;
